### PR TITLE
Shadows Enhancement

### PR DESCRIPTION
**Pull request by sebavan**
Add Close ESM Shadow
Add Kernel Blur Support
Fix Depth Compute Maths
Move Heavy Compute to Vertex
Optimize PostProcess chain
Authorize DepthValues per Lights

### CODE REVIEW

**Comment by deltakosh** [comment]
That's what I call a PR;)

File: src/Lights/Shadows/babylon.shadowGenerator.ts [hunk 1]
```diff
@@ -193,165 +308,177 @@
...
-                var mesh = subMesh.getRenderingMesh();
-                var scene = this._scene;
-                var engine = scene.getEngine();
+        private _initializeBlurRTTAndPostProcesses(): void {
+            var engine = this._scene.getEngine();
+            var targetSize = this._mapSize / this.blurScale;
 
-                // Culling
-                engine.setState(subMesh.getMaterial().backFaceCulling);
+            if (!this.useKernelBlur || this.blurScale !== 1.0) {
```
Comment (deltakosh) [inline src/Lights/Shadows/babylon.shadowGenerator.ts]: I would have  say !this.useKernelBlur && this.blurScale !== 1.0

File: src/Lights/babylon.directionalLight.ts [hunk 2]
```diff
@@ -163,5 +164,21 @@ module BABYLON {
             this._uniformBuffer.updateFloat4("vLightData", this.direction.x, this.direction.y, this.direction.z, 1, lightIndex);
             return this;
         }
+
+        /**
+         * Gets the minZ used for shadow according to both the scene and the light.
```
Comment (deltakosh) [inline src/Lights/babylon.directionalLight.ts]: perhaps a small comment here to explain const values

File: src/Shaders/ShadersInclude/shadowsFragmentFunctions.fx [hunk 3]
```diff
@@ -32,12 +31,12 @@
...
 	}
 
-	float computeShadowWithPCFCube(vec3 lightPosition, samplerCube shadowSampler, float mapSize, float darkness)
+	float computeShadowWithPCFCube(vec3 lightPosition, samplerCube shadowSampler, float mapSize, float darkness, vec2 depthValues)
 	{
 		vec3 directionToLight = vPositionW - lightPosition;
 		float depth = length(directionToLight);
-
-		depth = (depth - depthValues.x) / (depthValues.y - depthValues.x);
+		//depth = (depth - depthValues.x) / (depthValues.y - depthValues.x);
```
Comment (deltakosh) [inline src/Shaders/ShadersInclude/shadowsFragmentFunctions.fx]: could be removed

File: src/Shaders/ShadersInclude/shadowsFragmentFunctions.fx [hunk 4]
```diff
@@ -68,11 +67,12 @@
 		return  min(1.0, visibility + darkness);
 	}
 
-	float computeShadowWithESMCube(vec3 lightPosition, samplerCube shadowSampler, float darkness, float depthScale)
+	float computeShadowWithESMCube(vec3 lightPosition, samplerCube shadowSampler, float darkness, float depthScale, vec2 depthValues)
 	{
 		vec3 directionToLight = vPositionW - lightPosition;
 		float depth = length(directionToLight);
-		depth = (depth - depthValues.x) / (depthValues.y - depthValues.x);
+		//depth = (depth - depthValues.x) / (depthValues.y - depthValues.x);
```
Comment (deltakosh) [inline src/Shaders/ShadersInclude/shadowsFragmentFunctions.fx]: remove comments

File: src/Shaders/shadowMap.fragment.fx [hunk 5]
```diff
@@ -32,21 +28,10 @@ void main(void)
...
-	float depth = vPosition.z / vPosition.w;
-	depth = depth * 0.5 + 0.5;	
-#endif
-
-	depth += biasAndScale.x;
+	float depth = vDepthMetric;
 
 #ifdef ESM
-	depth = exp(-min(87., biasAndScale.y * depth));
+	depth = clamp(exp(-min(87., biasAndScale.y * depth)), 0., 1.);
```
Comment (deltakosh) [inline src/Shaders/shadowMap.fragment.fx]: the famous clamp :D

File: src/Lights/Shadows/babylon.shadowGenerator.ts (hunk 1 above)
Comment (sebavan) [inline src/Lights/Shadows/babylon.shadowGenerator.ts]: nope, still need the second map in box blur :-)

**Pull request merged**